src/features/textEditor/components/tipTapNodes/notionBlock/plugins/blockBrowsingPlugin.ts

```diff
@@ -0,0 +1,393 @@
+import { ReplaceAroundStep } from '@tiptap/pm/transform';
+import { Decoration, DecorationSet } from '@tiptap/pm/view';
+
+import { addUnindentSteps } from '../commands/unindent';
+import { NotionBlockNode } from '../NotionBlockNode';
+import { BlockSelection } from '../selection/BlockSelection';
+import { addIndentBlockSteps } from '../utils/addIndentBlockSteps';
+import { collapsibleArrowsPluginKey } from './collapsibleArrowPlugin';
+import { hideHandlePluginKey } from './hideHandlePlugin';
+
+interface BlockBrowsingPluginOnState {
+  enabled: true;
+  decorations: DecorationSet;
+}
+interface BlockBrowsingPluginOffState {
+  enabled: false;
+}
+
+type BlockBrowsingPluginState = BlockBrowsingPluginOnState | BlockBrowsingPluginOffState;
+
+export const blockBrowsingPluginKey = new PluginKey<BlockBrowsingPluginState>('blockBrowsing');
+export const blockBrowsingPlugin = new Plugin<BlockBrowsingPluginState>({
+  key: blockBrowsingPluginKey,
+  state: {
+    init: () => ({ enabled: false }),
+    apply: (tr) => {
+      const { selection } = tr;
+      if (!(selection instanceof BlockSelection)) {
+        return { enabled: false };
+      }
+      const decorations = DecorationSet.create(
+        tr.doc,
+        selection.selectedBlocksPos.map(({ from, to }) => Decoration.node(from, to, { class: 'selected' })),
+      );
+
+      return { enabled: true, decorations };
+    },
+  },
+  filterTransaction: (tr, state) => {
+    const pluginState = blockBrowsingPluginKey.getState(state);
+    if (!pluginState?.enabled) {
+      return true;
+    }
+    // Block every transaction except pointer events, collapsing/uncollapsing events, events dispatched by the plugin and events from hideHandlePlugin
+    return (
+      tr.docChanged ||
+      !!tr.getMeta('pointer') ||
+      !!tr.getMeta(blockBrowsingPluginKey) ||
+      tr.getMeta(collapsibleArrowsPluginKey) !== undefined ||
+      tr.getMeta(hideHandlePluginKey) !== undefined
+    );
+  },
+  props: {
+    decorations(state) {
+      const pluginState = this.getState(state);
+      if (pluginState?.enabled) {
+        return pluginState.decorations;
+      }
+    },
+    handleKeyDown(view, event) {
+      const state = this.getState(view.state);
+      const { selection, tr } = view.state;
+
+      if (!state?.enabled) {
+        if (event.code === 'Escape') {
+          tr.setSelection(new BlockSelection(selection.$anchor, selection.$head));
+          view.dispatch(tr);
+        }
+        return;
+      }
+
+      // This is not suppsoed to happen
+      if (!(selection instanceof BlockSelection)) {
+        console.error('Block browsing is enabled but no block is selected');
+        return;
+      }
+
+      const selectedBlock = view.state.doc.nodeAt(selection.head)!;
+      tr.setMeta(blockBrowsingPluginKey, true);
+      const depthDifference = selection.$from.depth - selection.$to.depth;
+      const endPos = selection.actualTo; // end of selection
+
+      switch (event.code) {
```

Comment:
nit: I see that this is the core logic for the block navigation. Can you make the code for each case in this switch own function?